### DISCUSSION TITLE
Use ExternalTaskSensor with SmartSensor

### DISCUSSION THREAD

**Discussion by ciancolo** [apache/airflow#13949] [Q&A]
Hello all,
Recently I updated my Airflow installation to the new version 2.0. One one of the most interesting features for my work is the smart sensor, so I am trying to add my old sensors to smart sensors for now without success.

To convert my sensors to smartsensor way I followed the guide in Airflow official documentation ([here](http://airflow.apache.org/docs/apache-airflow/stable/smart-sensor.html)). Here my configuration:
```
[smart_sensor]
use_smart_sensor = true
shard_code_upper_limit = 10000

# Users can change the following config based on their requirements
shards = 1
sensors_enabled = NamedHivePartitionSensor, MetastorePartitionSensor, ExternalTaskSensor
```

For my work, I use the ExternalTaskSensor. Here an example of a sensor:

```
DAG_DEFAULT_ARGS = {
    'owner': 'airflow',
    'email': ['myemail@bo.com'],
    'email_on_failure': True,
    'email_on_retry': False,
    'depends_on_past': False,
    'retry_delay': timedelta(minutes=15),
}

dag = DAG(
    dag_id="test_sensor",
    schedule_interval="0 8 * * *",
    start_date=datetime(2021, 1, 26),
    concurrency=1,
    catchup=False,
    default_args=DAG_DEFAULT_ARGS
)

SensorReadyForTest = ExternalTaskSensor(
    task_id='my_sensor_task',
    external_dag_id='my_external_dag',
    external_task_id='my_external_task',
    allowed_states=['success'],
    execution_delta=timedelta(seconds=0),
    poke_interval=60,
    mode='reschedule',
    timeout=60 * 60 * 2,
    dag=dag)

SensorReadyForTest

```
But the sensor is perfomed in old way. Here the log output of the run:

```
[2021-01-28 10:26:45,594] {taskinstance.py:1250} WARNING - Failed to register in sensor service.Continue to run task in non smart sensor mode.
[2021-01-28 10:26:45,594] {taskinstance.py:1252} ERROR - 'NoneType' object is not iterable
Traceback (most recent call last):
  File "/opt/venvs/airflow/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 1247, in _prepare_and_execute_task_with_callbacks
    registered = task_copy.register_in_sensor_service(self, context)
  File "/opt/venvs/airflow/lib/python3.7/site-packages/airflow/sensors/base.py", line 160, in register_in_sensor_service
    poke_context = self.get_poke_context(context)
  File "/opt/venvs/airflow/lib/python3.7/site-packages/airflow/sensors/base.py", line 178, in get_poke_context
    result = {key: getattr(self, key, None) for key in poke_context_fields}
TypeError: 'NoneType' object is not iterable
```
The sensor works fine in the old version but I am not able to run it in the new way. 
Did I miss some configurations or the ExternalTaskSensor is not compatible with SmartSensor?
I checked the code of ExternalTaskSensor and BaseSensor and I am not sure that this type of sensor is compatible, for example in NamedHivePartitionSensor is defined the variable poke_context_fields, while in ExternalTaskSensor is not.

Could anyone help me?

Thank you

**Comment by stevenzhang-support** [comment]
Hello I have a user who reported the same issue when using smart sensor feature with ExternalTaskSensor. Please the Product team kindly reply this question above

**Reply by potiuk** [reply]
FYI. There is no product team. There are indivduals who contribute to the project (~ 1900 contributors), There are maintainers by they are mostly volunteers. This does not work the way commercial product works. People help here on a best effort case here. If you or your users need commercial support, there are companies that can offer that but here you will see mostly people helping "when they can".

I think you should not use Smart Sensors. They are being deprecated and will be removed in 2.4. It was an early-access/preview feature that we always warned that will be removed. You should migrate to Airflow 2.2 and use Deferrable Operators instead

Here info about Smart Sensors: https://airflow.apache.org/docs/apache-airflow/stable/concepts/smart-sensors.html?highlight=smart%20sensors

Deferrable Operators here: https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html
